### DISCUSSION TITLE
Are there any best practices for migrating from Chevrotain to Langium?

### DISCUSSION THREAD

**Discussion by MrKou47** [eclipse-langium/langium#1331] [General]
Hi guys, I have already created my own language using Chevrotain. However, I recently discovered Langium, which offers better Language Server Protocol (LSP) support. I would like to try using Langium to rewrite my language. Is there any manual or documentation available for migrating from Chevrotain to Langium?

**Comment by spoenemann** [comment]
I'm not aware of any migration guide, but the best place to start might be the [Grammar Language documentation](https://langium.org/docs/grammar-language/). There's also a [Playground](https://langium.org/playground/) to try it right on the website.

**Comment by msujew** [comment]
While @spoenemann is correct, and there is no manual or guide for migration from chevrotain, I can offer a few tips for that:

1. If you're using the railroad syntax diagrams generated for Chevrotain (see [here](https://chevrotain.io/docs/guide/generating_syntax_diagrams.html)), you can use them as a guideline to migrate your language to Langium. Langium offers these railroad diagrams as well and you can get a live preview of them in the [vscode extension](https://marketplace.visualstudio.com/items?itemName=langium.langium-vscode).
2. Langium doesn't support syntactic or complicated semantic predicates. If your language relies on that, you might need to refactor some of it.
3. Please note that Langium works best if your language is making use of the [cross references feature](https://langium.org/docs/grammar-language/#cross-references). This isn't something that's supported by Chevrotain, but Langium needs this information to offer most of the LSP features (like jumping to declarations, finding references, code completion, etc.) by default. You can reimplement those though, if necessary.

**Reply by MrKou47** [reply]
Thanks for your comprehensive answer! I'll give it a try :)
